### DISCUSSION TITLE
best way to pull in a file from a sibling directory?

### DISCUSSION THREAD

**Discussion by greglearns** [ipetkov/crane#395] [General]
My `flake.nix` file is right next to the rust `build.rs` file, and build.rs needs to load a file `../spec/service.proto`. All of these files are I've read the documentation https://crane.dev/source-filtering.html but I'm relatively new to nix, I'm not sure how to make this work, and I've tried a few times to make it work and I can't figure it out. I have this directory structure:
|root
|.git <--- all of the files below have been `git add`ed and so the flake should be able to find them.
|-- rust
  |-- flake.nix
  |-- build.rs  <-- this needs to read the ../spec/service.proto file.
  |-- src/
|-- spec/service.proto

Any suggestions?

P.S. Thank you for crane!

**Comment by ipetkov** [comment]
Hi @greglearns ! Could you share your `flake.nix` so I can see how the build is configured?

Two possible situations come to mind:
1. Is `spec/service.proto` staged with git at all? (e.g. have you run `git add -N spec/service.proto`?) Flakes can only see files which are staged (or committed) with git and aren't listed in `.gitignore`
2. You mentioned you've looked at source filtering. One quick thing to check is setting `src = ./.;` which will not do any additional source filtering (besides ignoring untracked files). If the build succeeds with this then its a matter of tweaking the source filter code, otherwise it might be something else!

**Comment by greglearns** [comment]
https://gist.github.com/greglearns/a8f63f33040248411f5ef281dde51e33

1. `spec/service.proto` is staged in git. The git repo is at the root level, with `./rust/flake.nix` and `./spec/service.proto` being subdirectories under the root directory where the `.git` dir lives.

2. I think the problem is that the flake is in the `./rust/` directory, so `src = ./.` is only looking in the `./rust` directory, but the proto file is in the `./spec` directory.

3. If I copy `./spec/service.proto` into the `./rust` directory as `./spec/copied-service.proto`, and then change the `./rust/build.rs` file to point to the copied version at `./rust/copied-service.proto` file, then run `nix build` then it all works. But. I can't do that since the `spec/service.proto` is shared by other projects and can't be moved.

**Comment by ipetkov** [comment]
Ah I see! I had misunderstood the original description and didn't realize that `flake.nix` was inside of the `rust` directory.

As far as I understand, flakes do not allow accessing files in parent directories (only sub-directories relative to their location) so there's two approaches which come to mind for resolving this:
1. The simplest is to move `flake.nix` to the root of the repo. Newer versions of `nix` will automatically look for a `flake.nix` in a parent directory so you can theoretically still run things like `nix build` etc. inside of `./rust`
2. It may be possible to add a symlink in `./rust/spec -> ../spec` and get that to work, assuming that when the flake is evaluated it performs a deep copy of the symlink's contents in the store instead of copying the link itself. I haven't actually tested this so if it doesn't work, option 1 above would be my go to!

Hope this helps!
